File: sample-chat/sample-chat/Services/UsersService.m
//
//  UsersService.m
//  sample-chat
//
//  Created by Anton Sokolchenko on 5/26/15.
//  Copyright (c) 2015 Igor Khomenko. All rights reserved.
//

#import "UsersService.h"
#import "ServicesManager.h"

@implementation UsersService

- (instancetype)initWithContactListService:(QMContactListService *)contactListService {
	self = [super init];
	if( self ) {
		_contactListService = contactListService;
	}
	return self;
}

- (void)cachedUsersWithCompletion:(void(^)(NSArray *users))completion {
	// check memory storage
	NSArray *memoryUsers = [self.contactListService.usersMemoryStorage usersSortedByKey:@"fullName" ascending:YES];
	if (memoryUsers != nil && memoryUsers.count != 0) {
        if (completion) {
            completion(memoryUsers);
        }
        
        return;
	}
	
	// check CoreData storage
	[QMContactListCache.instance usersSortedBy:@"fullName" ascending:YES completion:^(NSArray *users) {
        if (completion) {
            completion(users);
        }
	}];
}

- (void)downloadLatestUsersWithSuccessBlock:(void(^)(NSArray *latestUsers))successBlock errorBlock:(void(^)(QBResponse *response))errorBlock {
	__weak __typeof(self)weakSelf = self;
	
    /**
     *  Different users are taken depending on environment.
     */
    NSString* environment = nil;
#if DEV
    environment = @"dev";
#endif
    
#if QA
    environment = @"qbqa";
#endif
    
#if RELEASE
    environment = @"release";
#endif
    
    [QBRequest usersWithTags:@[environment] successBlock:^(QBResponse *response, QBGeneralResponsePage *page, NSArray *users) {
        __typeof(self) strongSelf = weakSelf;
        
        NSMutableArray* mutableUsers = [users mutableCopy];
		[mutableUsers sortUsingComparator:^NSComparisonResult(QBUUser *obj1, QBUUser *obj2) {
			return [obj1.login compare:obj2.login options:NSNumericSearch];
		}];

		[strongSelf.contactListService.usersMemoryStorage addUsers:users];
		[QMContactListCache.instance insertOrUpdateUsers:[mutableUsers copy] completion:nil];

		if (successBlock != nil) {
			successBlock([mutableUsers copy]);
		}
    } errorBlock:^(QBResponse *response) {
		if (errorBlock != nil) {
			errorBlock(response);
		}
		NSLog(@"error: %@", response.error.error);
    }];
}

- (NSArray *)usersWithoutCurrentUser {
	NSArray *usersWithoutCurrentUser = [self.contactListService.usersMemoryStorage unsortedUsers];
	if ([QBSession currentSession].currentUser) {
		usersWithoutCurrentUser = [usersWithoutCurrentUser filteredArrayUsingPredicate:[NSPredicate predicateWithBlock:^BOOL(id evaluatedObject, NSDictionary *bindings) {
			QBUUser *user = (QBUUser *)evaluatedObject;
			return user.ID != [QBSession currentSession].currentUser.ID;
		}]];
	}
	
	return usersWithoutCurrentUser;
}

- (NSArray *)idsWithUsers:(NSArray *)users {
	return [users valueForKeyPath:@"@distinctUnionOfObjects.ID"];
}

- (void)retrieveUsersWithIDs:(NSArray *)usersIDs completion:(void(^)(QBResponse *response, QBGeneralResponsePage *page, NSArray *users))completion {
	[self.contactListService retrieveUsersWithIDs:usersIDs forceDownload:NO completion:completion];
}

@end

File: sample-chat-swift/sample-chat-swift/Classes/Services/UsersService.swift
//
//  UsersService.swift
//  sample-chat-swift
//
//  Created by Injoit on 7/17/15.
//  Copyright (c) 2015 quickblox. All rights reserved.
//

import Foundation

/**
*  Service responsible for working with users. 
*/
class UsersService: NSObject {
    
    var contactListService : QMContactListService!
    
    var colors = [
        UIColor(red: 0.992, green:0.510, blue:0.035, alpha:1.000),
        UIColor(red: 0.039, green:0.376, blue:1.000, alpha:1.000),
        UIColor(red: 0.984, green:0.000, blue:0.498, alpha:1.000),
        UIColor(red: 0.204, green:0.644, blue:0.251, alpha:1.000),
        UIColor(red: 0.580, green:0.012, blue:0.580, alpha:1.000),
        UIColor(red: 0.396, green:0.580, blue:0.773, alpha:1.000),
        UIColor(red: 0.765, green:0.000, blue:0.086, alpha:1.000),
        UIColor.redColor(),
        UIColor(red: 0.786, green:0.706, blue:0.000, alpha:1.000),
        UIColor(red: 0.740, green:0.624, blue:0.797, alpha:1.000)
    ]
    
    init(contactListService: QMContactListService) {
        super.init()
        self.contactListService = contactListService
    }
    
    
    func cachedUser(completion:(([QBUUser]) -> Void)?) {
        
        weak var weakSelf = self
        
        // check memory storage
        
        let memoryUsers = self.contactListService!.usersMemoryStorage.usersSortedByKey("fullName", ascending: true)
        
        if (memoryUsers != nil && memoryUsers.count > 0) {
            
            let sortedUsers = UsersService.sortedUsers(memoryUsers as! [QBUUser])
            
            completion?(sortedUsers)
            
            return
        }
    
        // check persistent storage
        
        QMContactListCache.instance().usersSortedBy("login", ascending: true) { (users: [AnyObject]!) -> Void in
            
            let sortedUsers = UsersService.sortedUsers(users as! [QBUUser])
            
            weakSelf?.contactListService!.usersMemoryStorage.addUsers(sortedUsers)
            
            completion?(sortedUsers)
        }
    }
    
    func downloadLatestUsers(success:(([QBUUser]) -> Void)?, error:((QBResponse) -> Void)?) {
        
        weak var weakSelf = self
        
        let enviroment = Constants.QB_USERS_ENVIROMENT
        
        QBRequest.usersWithTags([enviroment], successBlock: { (response: QBResponse!, page: QBGeneralResponsePage?, users: [QBUUser]?) -> Void in
            
            let sortedUsers = UsersService.sortedUsers(users!)

            weakSelf?.contactListService!.usersMemoryStorage.addUsers(sortedUsers)
            QMContactListCache.instance().insertOrUpdateUsers(sortedUsers, completion: nil)
            
            success?(sortedUsers)
            
            }) { (response: QBResponse!) -> Void in
            
                error?(response)
                
                NSLog("error: %@", response)
        }
    }
    
    func user(userID: UInt) -> QBUUser? {
    
        return self.contactListService.usersMemoryStorage.userWithID(userID)
    }
    
    func users() -> [QBUUser]? {
        return self.contactListService.usersMemoryStorage.unsortedUsers() as? [QBUUser]
    }
    
    func users(withoutUser user: QBUUser) -> [QBUUser]? {
        
        return (self.contactListService.usersMemoryStorage.unsortedUsers() as! [QBUUser]).filter({$0.ID != user.ID})
    }
    
    static func sortedUsers(unsortedUsers: [QBUUser]) -> [QBUUser] {
        
        let sortedUsers = unsortedUsers.sort({ (user1, user2) -> Bool in
            return (user1.login! as NSString).compare(user2.login!, options:NSStringCompareOptions.NumericSearch) == NSComparisonResult.OrderedAscending
            })
        
        return sortedUsers
    }
    
    func color(forUser user:QBUUser) -> UIColor {
        
        let userIndex = (self.users()!).indexOf(self.user(user.ID)!)
        
        if userIndex < self.colors.count {
            return self.colors[userIndex!]
        } else {
            return UIColor.blackColor()
        }
    }
}